src/Chapter3.hs

```diff
@@ -452,14 +475,40 @@ concept of product types and sum types is called __Algebraic Data Type__. They
 allow you to model your domain precisely, make illegal states unrepresentable
 and provide more flexibility when working with data types.
 -}
+data Hero 
+    = DeadWarrior 
+    | Warrior
+      { warriorHealth :: Int
+      , warriorAttack :: Int
+      , warriorGold :: Int
+      } deriving (Eq, Show)
```

Comment:
To have records inside a sum type is an anti-pattern in Haskell. And it is not recommended to use it as it creates partial functions. As we mentioned, record named fields create the functions visible for the whole module. In this case it will also create such functions. E.g.:
```haskell
warriorHealth :: Hero -> Int
```

The problem here is that this function can fail in the runtime. If you try to `warriorHealth DeadWarrior` it will fail.

You have several options here:
1. Separate that into the normal record and use that in the sum type:
    ```haskell
    data Warrior = Warrior
        { warriorHealth :: Int
        , warriorAttack :: Int
        , warriorGold :: Int
        } deriving (Eq, Show)

    data Hero 
        = DeadWarrior 
        | AliveWarrior Warrior
    ```
2. Use unnamed fields in sum type:
    ```haskell
    data Hero
        = DeadWarrior
        | Warrior Int Int Int
    ```
    which is not that convenient in this case, as you have three `Int`s.
3. You can have just record type, and find out is warrior dead or not with a function:
    ```haskell
    data Warrior = Warrior
        { warriorHealth :: Int
        , warriorAttack :: Int
        , warriorGold :: Int
        } deriving (Eq, Show)

    isDead :: Warrior -> Bool
    isDead w = warriorHealth <= 0
    ```

Reply:
Oh, thank you! I have this problem everywhere 😅 I guess I'll stick with the first option this time.